### ISSUE TITLE
400 error transferring to Clockify

### ISSUE THREAD

**Issue by TheSpyder** [mikerourke/transfermyti.me#40]
I just imported some spreadsheet data into a brand new Toggl account, the time entries have no project or description. I've been using Clockify for about a month and decided to try importing via Toggl rather than wait for Clockify to finish their import feature.

When attempting to transfer time entries it triggers a 400 error from clockify (the page reports transfer complete but there's a 400 error notification).

Digging in to the developer tools, here's the request data that triggers it:
Request URL: `https://api.clockify.me/api/v1/workspaces/<some numbers>/time-entries`
Request payload:
`{"start":"2020-07-01T22:30:00.000Z","billable":false,"description":"","end":"2020-07-02T05:45:00.000Z"}`

I'm in Australia (GMT+10) and those times look correct. I'm not sure why clockify is rejecting them.

**Comment by mikerourke** [comment]
I’ll take a look. There may have been changes to the Clockify API that I’m not aware of. Let me do some digging and get back to you.

**Comment by mikerourke** [comment]
@TheSpyder just to give you an update, it's a busy week at work so I won't be able to troubleshoot the issue until this weekend. I'll let you know as soon as I find anything. I apologize for the delay.

**Comment by TheSpyder** [comment]
Thanks for letting me know! There's no rush, I'm just using it as a convenient way to import data into Clockify without learning the API myself 😂

**Comment by mikerourke** [comment]
@TheSpyder I apologize for getting back to you so late. It's been a...let's say "tense" week in the US. I was able to create a time entry on Clockify matching your payload, but after re-reading your issue, I noticed that the Request URL (`https://api.clockify.me/api/v1/workspaces//time-entries`) is missing a workspace ID between `workspaces/` and `/time-entries`). Did you create a workspace on Clockify prior to the transfer? If not, I think there may be a bug in my code that doesn't update the workspace ID after creating it.

If you haven't already, create the workspace on Clockify first, then try the transfer again and it should work. Also, I remember Clockify was having issues right around the time you were trying to perform the transfer, so that might also be a contributing factor.

I'm going to dig into my code to make sure I'm not missing anything. I'm wondering if up to this point, all the users have just been creating a workspace on Clockify before trying the transfer. I'll let you know what I find. Thanks for your patience.

**Comment by TheSpyder** [comment]
Oh I forgot to put backticks around the URL! Sorry, I replaced the workspace ID with what github interpreted as a HTML tag figuring it might be private information. I've corrected the description.

I'll give it another go today and let you know.

**Comment by mikerourke** [comment]
Thanks! I'm going through and updating the Clockify API calls now. They were transitioning from an older API to a new one and it looks like they completely deprecated the old API. They removed the ability to create workspaces, so it looks like they need to be created now before even using the transfer tool. I'm testing out transfers now to make sure I didn't break anything, so I'm hoping I don't run into the same issue you had and it was because of a Clockify service outage. 🤞

**Comment by TheSpyder** [comment]
In the time since logging this issue I've had a bit more experience with fetch, so I know that if things aren't perfect the browser generally ignores any response text.

Thanks to copy as curl, I can tell you the actual error returned by the API is
`{"message":"Time entry couldn't be created. Project is either required field or given project is archived. Please change your input or update your workspace settings.","code":501}`

**Comment by TheSpyder** [comment]
Interesting, based on that error I bulk edited my toggl time entries to use a project with the same name as the project I've created in clockify.

Now the import tool thinks all of my time entries have already been imported. They don't have a task or description, and the only time entries that exist in clockify are for later days than the toggl entries.

**Comment by mikerourke** [comment]
Could you try using my transfer tool to bulk-delete your Toggl data and import them again? I pushed an update last night that addressed the Clockify API deprecation. I'm wondering if a payload or endpoint changed and I missed it. I might just go through everything again with a fine tooth comb.

**Comment by TheSpyder** [comment]
Unfortunately it's given me the same error. As before I tried both fresh data with no project, which failed, and I confirmed the response error message was the same. I then bulk-edited the Toggl entries to give them a project matching the name of my Clockify project. All that did was make the tool think every time entry had already been imported (even though no Clockify time entry exists on those dates).

What can I do to help track this down? Make a new workspace? Make a new test account? I can perhaps send you some of the data I imported but I'm not sure that will help if it's a clockify issue.

**Comment by TheSpyder** [comment]
I started reading the clockify api docs and comparing what's returned in the chrome network traces from your API calls. Looking at my `/workspaces` response, it has `"forceProjects":true`. That probably explains the error! I have no idea how that flag was set, I can't see anything in my workspace settings.

So now the question is why when I set a project in Toggl does the app mark all existing entries as already imported 🤔

**Comment by mikerourke** [comment]
Good catch! I didn't even know about that `forceProjects` setting. In terms of next steps, could you send me the CSV that you imported into Toggl? You can delete like all but 3 entries and change/remove any sensitive data. I'd prefer to try doing the exact same thing you tried to see if I can reproduce the error in my test accounts. The time entry duplication check is _not_ very reliable, so I'm not surprised they're being marked as duplicates. I would recommend just deleting everything from Clockify using the app before trying the transfer again.

**Comment by TheSpyder** [comment]
The data is quite simple, here's a weeks worth with the email removed:
[Toggl Import.csv.zip](https://github.com/mikerourke/transfermyti.me/files/5527239/Toggl.Import.csv.zip)

I can't delete everything from clockify, it contains my current data, I'm just trying to import time entries from the manual spreadsheet system I used before. I could perhaps create a new workspace to test with? I guess I could also transfer from clockify to toggl as a backup 🤭 😂 

Is there a way to override the duplication check? I wouldn't mind force-importing everything since there are only 52 entries and I know they aren't in clockify.

**Comment by mikerourke** [comment]
You can test it out with a different workspace to make sure it works, but you can't move time entries between workspaces (to my knowledge), so that won't fix your problem. If you have a bunch of Clockify data already, the app lets you pick and choose what you want to delete, so you don't have to start from scratch. You can always hedge your bets and transfer your Clockify data to Toggl, but I deliberately built the delete functionality to error on the side of caution.

You can't disable the duplication check for time entries now, but it's a good idea to offer that (considering how crude my duplication check is). I started working on that tonight, but it might be another day or two until it's ready to use. I don't want to rush and break something else. If you don't mind waiting a couple of days, I can let you know as soon it's ready.

**Comment by TheSpyder** [comment]
Waiting for the override sounds like the easiest solution for me.

I am in no rush at all to import this data, it's safe in my spreadsheet and I won't need to generate a report based on it until next year, but I've pretty much given up on waiting for Clockify to implement import themselves 😅

**Comment by mikerourke** [comment]
I'll keep you posted. Thanks for your patience!

**Comment by TheSpyder** [comment]
On a hunch, I created a test clockify workspace, confirmed that I could import all toggl entries into it, then wiped it and created a single time entry in clockify. _This reproduced the issue where all entries were marked as already imported_.

I don't want to go spelunking through your code but my guess is it's related to timezones. I'm in GMT+10, which I have set in both apps, but looking at the API data toggl returns a timezone-based format whereas Clockify returns the UTC-based format. After discovering this I went back and added an actually duplicate time entry.

Toggl:
```
      "start": "2020-09-25T08:30:00+10:00",
      "end": "2020-09-25T17:15:00+10:00",
      "updated": "2020-11-11T17:35:09+10:00",
      "dur": 31500000,
```

Clockify:
```
    "timeInterval": {
      "start": "2020-09-24T22:30:00Z",
      "end": "2020-09-25T07:15:00Z",
      "duration": "PT8H45M"
    },
```

There are many factors with timezone conversion that could be triggering your duplicate entry check incorrectly. This single clockify entry renders all Toggl data as duplicate. Now that I know this, I'm happy to help set up some test accounts if you'd like to try addressing the duplication check instead of adding a way to disable it 🤔

**Comment by mikerourke** [comment]
I looked through my code and it looks like I'm creating `Date` objects from both values before comparing them, so those times should match. I'll definitely validate this while I'm working on the duplication check toggle. I think I might run into issues with all entries being marked as duplicated though. Right now, I'm looping through each source time entry, and within that loop, I'm looping through each target time entry (in our case, Toggl), and checking if the projects match, the descriptions match, and the difference in minutes between the start and stop <= 1. I can try reversing the loop order, but I suspect the reason this hasn't been an issue up until now is that most users wouldn't have 10 time entries with no project, no description, and the same start/stop time.

That being said, I think I'm going to just add the ability to disable the duplication check, but have it enabled by default. I'll work on wrapping up that feature so you can still do your transfer. I'm apprehensive about doing too much with the duplicate check logic because I'd rather have false positives than false negatives. There have been a few people that used the tool to transfer several thousand records from Toggl to Clockify and had to do the transfer incrementally because the servers couldn't handle it. Having a bunch of duplicate time entries because the duplication check was too lax would be frustrating. It's been a while since I touched this part of the code though, so I could be completely wrong. Thank you so much for digging into this! Working with dates and times tends to be a nightmare and I appreciate having someone from a completely different timezone to help with reproducing these issues.

I'm continuing to chip away at the duplication check in my off hours. You'll be the first to know when it's ready to test. Thanks again!

**Comment by TheSpyder** [comment]
Quite a lot of my entries have the same start time (the toggl data is overly simplified due to how I set up my spreadsheet, it's only one block of time per day) but none of them have the same start _date_.

Another data point: I wiped the test clockify workspace, imported a single time entry (the one I posted above) and when I re-ran the tool it marked all toggl data as duplicate. You're right, `new Date().getTime()` should return the correct value regardless of timezone, so maybe the bug lies somewhere else.

**Comment by mikerourke** [comment]
That is very strange. I ended up having to work late yesterday so I didn't put as much of a dent in the duplication check toggle as I wanted. I'm going to hedge my bets here and say it'll be ready to go by Monday. It would be cool if I stumbled across the duplication bug by accident and was able to fix it at the same time, but I'm skeptical about that actually happening. I'll let when toggle feature is deployed.

**Comment by mikerourke** [comment]
@TheSpyder, heads up, I was able to knock that duplication check toggle out tonight. I wasn't able to track down what was causing the check bug, but at least you can manually disable it for now. Give it a whirl and let me know if you have any issues.

**Comment by TheSpyder** [comment]
Awesome, thank you! I won’t have time until my evening but I will let you know.

**Comment by TheSpyder** [comment]
That worked 🎉 

I was fiddling around with the data and interestingly when I set my Toggl data to have a project suddenly the duplication check on my "test" workspace operated correctly; it was only data without a project that it marked as duplicate. I took a quick poke around the data returned by the Toggl API and the only difference I could see was that `pid` was `null` in the Toggl data that incorrectly triggers the duplication check.

On my real workspace I had the same issue as before but overriding the duplication check worked perfectly.

I then imported my older data to Toggl, which is in a new dedicated project, and transferring it worked without any extra hassle; none of the entries I wanted to import were flagged as duplicate. I didn't create the project in Clockify before I started, tool imported both the project and all the time entries successfully.

I am curious why my data caused these problems, it doesn't make a whole lot of sense, but for now my problem is definitely fixed. Thank you 😁
